Import API changes

cmark opened the pull request:
New content import Java API has been developed to provide developers the necessary tools for their plug-ins to easily validate and import content into Snow Owl Terminology Server.

cmark committed:
SO-4483: support `dryRun` in RF2 import Java/REST APIs

cmark committed:
[io] commit progress on new generic import API

cmark committed:
SO-4346: add CodeSystemURI based build methods

cmark committed:
[core] add hasChanges to CommitResult

Make commitTimestamp primitive.

cmark committed:
SO-4346: improve Import Defect API

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/io/ImportDefectAcceptor.java:
```diff
@@ -0,0 +1,142 @@
+		}
+		
+		public ImportDefectBuilder whenNaN(String value) {
+			return when(!CharMatcher.inRange('0', '9').matchesAllOf(value));
```
Are floating point dots/commas and negative signs allowed? Maybe E for powers of 10? 😃

cmark replied:
Let's keep it simple for the initial release, but this would be great to support.

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/io/ImportRequest.java:
```diff
@@ -0,0 +1,143 @@
+			doValidate(context, attachment, defectsAcceptor, context.service(IProgressMonitor.class));
+			
+			final List<ImportDefect> defects = defectsAcceptor.getDefects();
+			if (!CompareUtils.isEmpty(defects)) {
```
Informational and warning messages could be let through here...

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/io/ImportRequest.java:
```diff
@@ -0,0 +1,143 @@
+			} else {
+				final Set<ComponentURI> visitedComponents = Sets.newHashSet();
+				doImport(context, attachment, visitedComponents::add, context.service(IProgressMonitor.class));
+				return ImportResponse.success(visitedComponents);
```
...and then it would make sense to return them along with the components imported.

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/io/ImportRequest.java:
```diff
@@ -0,0 +1,143 @@
+//	}
+	
+	/**
+	 * Subclasses optionally provider validation functionality to verify the integrity of the attachment before proceeding to the actual import in {@link #doImport(File, IProgressMonitor)}.
```
```suggestion
	 * Subclasses can optionally provide validation functionality to verify the integrity of the attachment before proceeding to the actual import in {@link #doImport(File, IProgressMonitor)}.
```

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/io/ImportRequest.java:
```diff
@@ -0,0 +1,143 @@
+	 * @param attachment - the file attachment to validate
+	 * @param defectsAcceptor - the acceptor that collects {@link ImportDefect}s through a few helper methods
+	 * @param monitor - the monitor that can be used to track progress
+	 * @return a {@link List} of {@link ImportDefect} instances that can aid the user pinpoint and resolve the errors/issues
```
There is no return value, the import defect collecting object consumes all messages instead.

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/io/ImportRequest.java:
```diff
@@ -0,0 +1,143 @@
+	 * @param attachment - the attachment file to import
+	 * @param visitor - visitor that accepts visited component URIs
+	 * @param monitor - the monitor that can be used to track progress
+	 * @return
```
No return value in this method either.

apeteri reviewed (approved):
👍

cmark committed:
SO-4346: add Serializable to ImportDefect

cmark committed:
SO-4346: return warn/info defects as part of the success result

cmark committed:
SO-4346: log error when request return null response body

cmark committed:
SO-4346: clean up in ImportRequest

cmark committed:
Update core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/io/ImportRequest.java

Co-authored-by: András Péteri <apeteri@b2international.com>

cmark committed:
SO-4346: fix log statement substitution

cmark committed:
SO-4346: fix unused format argument

cmark committed:
SO-4346: support optionalService up to the root context

cmark committed:
SO-4346: fall back to Locks.lockContext if parentLockContext not defined

...in RepositoryTransactionContext (requires outer wrapping of Locks)

For any other inner logic that uses `TransactionContext` interface with
a custom made lock should still delegate its lockContext to the
`TransactionContext.commit(...)` calls.

cmark committed:
SO-4346: move finished log message before sending back response

cmark committed:
Merge branch '7.x' into improvements/import-api-refactor

cmark committed:
Merge branch '7.x' into improvements/import-api-refactor

apeteri committed:
Merge remote-tracking branch 'origin/7.x' into
improvements/import-api-refactor

Conflicts:
	core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/repository/RepositoryTransactionContext.java

apeteri committed:
SO-4346: Use ImportDefectBuilder for RF2 content validation

apeteri committed:
SO-4346: Restrict maximum collected defects in ImportDefectAcceptor

apeteri committed:
SO-4346: Use generic ImportResponse in SnomedRf2ImportRequest

- Update Rf2GlobalValidator to use an import defect collecting instance
- Update Rf2ValidationIssueReporter to collect defects by source file
- Collect visited component URIs in Rf2EffectiveTimeSlice
- Update SnomedCommand after API change
- Set ImportStatus based on ImportResponse in SnomedRf2ImportService

apeteri committed:
SO-4346: Validate effective time slices in reverse chronological order

apeteri committed:
SO-4346: Create a copy of the HTreeMap's key set

One-time iteration of the key set is fast, but single "contains" checks
are comparatively very slow.

apeteri committed:
SO-4346: Register only concepts and reference sets in visited components

apeteri committed:
SO-4346: Treat empty "when" conditions as defects in...

...ImportDefectBuilder; the "error", "warn", "info" shortcuts in
ImportDefectAcceptor won't register a defect otherwise.

apeteri committed:
SO-4346: Initialize defect location field with an empty string

apeteri committed:
Merge remote-tracking branch 'origin/7.x' into improvements/import-api-refactor

cmark commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/io/ImportDefectAcceptor.java:
```diff
@@ -0,0 +1,151 @@
+				defects.add(new ImportDefect(file, location, message, type));
+
+				// Remove the earliest defect on overflow
+				if (defects.size() > DEFAULT_MAX_DEFECTS) {
```
I suggest returning all defects from the API and limiting visualization on user interfaces instead.

cmark reviewed (commented):
LGTM!

apeteri committed:
Revert "SO-4346: Restrict maximum collected defects in ImportDefectAcceptor"

This reverts commit 082c123fdc78ec134d6e363bee67db3baacd7211.

apeteri committed:
SO-4346: Remove restriction on defects from Rf2ValidationIssueReporter
